### PR TITLE
fix invalidations related to `ismutable`

### PR DESCRIPTION
**Pull request by ranocha**
Related to #52134. It would be nice if the underlying inference issue was fixed but this seems to be a hotfix for now.

I have seen this inference problem occurring in Julia v1.9, v1.10, and current `master`. For example, on Julia v1.9.3, I get

```julia

julia> code_warntype(ismutable, (Function,))
MethodInstance for ismutable(::Function)
  from ismutable(x) @ Base reflection.jl:521
Arguments
  #self#::Core.Const(ismutable)
  x::Function
Body::Any
1 ─      nothing
│        nothing
│   %3 = Base.typeof(x)::Type{<:Function}
│   %4 = Base.getproperty(%3, :name)::Any
│   %5 = Base.getproperty(%4, :flags)::Any
│   %6 = (%5 & 0x02)::Any
│   %7 = (%6 == 0x02)::Any
└──      return %7
```

This causes some invalidations when `using OrdinaryDiffEq`.

### CODE REVIEW

**Review by oscardssmith — APPROVED** [review]

File: base/reflection.jl [hunk 1]
```diff
@@ -627,7 +627,9 @@ true
 !!! compat "Julia 1.5"
     This function requires at least Julia 1.5.
 """
-ismutable(@nospecialize(x)) = (@_total_meta; typeof(x).name.flags & 0x2 == 0x2)
+ismutable(@nospecialize(x)) = (@_total_meta; (typeof(x).name::Core.TypeName).flags & 0x2 == 0x2)
```
Comment (vtjnash) [inline base/reflection.jl:630]: perhaps this?

```suggestion
ismutable(@nospecialize(x)) = (@_total_meta; (Base.inferencebarrier(typeof(x))::DataType).name.flags & 0x2 == 0x2)
```

(though we really need to work on teaching inference to do this itself)

Comment (ranocha) [reply base/reflection.jl:630]: I get 

```
error during bootstrap:
LoadError(at "compiler/compiler.jl" line 3: LoadError(at "compiler/validation.jl" line 4: UndefVarError(var=:Base, scope=Core.Compiler)))
```

when doing this locally and running `make`.

Comment (ranocha) [reply base/reflection.jl]: I implemented the corrected version in [3f5436c](https://github.com/JuliaLang/julia/pull/52170/commits/3f5436c633943ceb0f101022a6b3952e3fb3f3f6)

**Comment by ranocha** [comment]
Bump

**Comment by KristofferC** [comment]
Tests fail. Did they fail with the earlier version as well?

**Comment by ranocha** [comment]
CI was green with the first commit in this PR (before applying the suggestion):

Screenshot 2023-11-22 at 08 50 41

**Comment by KristofferC** [comment]
Maybe revert to that then?

**Comment by ranocha** [comment]
> Maybe revert to that then?

Done in ab6db687526c817e2214f30582a5d3cfbc409130

**Comment by ranocha** [comment]
Test failures like

```
floatfuncs                                        (8) |        started at 2023-11-22T08:45:13.525
      From worker 7:	ERROR: LoadError: UndefVarError: `@which` not defined in `Main`
```
(https://buildkite.com/julialang/julia-master/builds/30374#018bf613-7cf7-4ce1-b8b9-3d391bcb6c64/783-1040)

```
Error During Test at /cache/build/tester-amdci4-12/julialang/julia-master/julia-ab6db68752/share/julia/stdlib/v1.11/SparseArrays/test/ambiguous.jl:47
  Got exception outside of a @test
  UndefVarError: `test_project_toml_formatting` not defined in `Aqua`
```
(https://buildkite.com/julialang/julia-master/builds/30374#018bf613-7cf7-4ce1-b8b9-3d391bcb6c64/783-1141)

```
Error in testset SparseArrays/ambiguous:
Test Failed at /cache/build/tester-amdci4-12/julialang/julia-master/tmp/jl_aUFVj1/packages/Aqua/jBmtw/src/deps_compat.jl:59
  Expression: isempty(result)
   Evaluated: isempty(Base.PkgId[Libdl [8f399da3-3557-5675-b5ff-fb832c97cbdb], LinearAlgebra [37e2e46d-f89d-539d-b4ee-838fcccc9c8e], Random [9a3f8284-a2c9-5f02-9a11-845980a1fd5c], Serialization [9e88b42a-f829-5b0c-bbe9-9e923198166b]])
```
(https://buildkite.com/julialang/julia-master/builds/30374#018bf613-7cf7-4ce1-b8b9-3d391bcb6c64/783-1654)

do not seem to be related to this PR, are they?

**Comment by ranocha** [comment]
CI passes

**Pull request merged**